rasa_core/featurizers.py

```diff
@@ -163,18 +136,425 @@ def encode(self, active_features, input_feature_map):
+        y = self._featurize_labels(trackers_as_actions, domain)
+
+        return (DialogueTrainingData(X, y, metadata),
+                true_lengths)
+
+    def prediction_states(self,
+                          trackers,  # type: List[DialogueStateTracker]
+                          domain  # type: Domain
+                          ):
+        # type: (...) -> List[List[Dict[Text, float]]]
+        raise NotImplementedError("Featurizer must have the capacity to "
+                                  "create feature vector")
+
+    def create_X(self,
+                 trackers,  # type: List[DialogueStateTracker]
+                 domain  # type: Domain
+                 ):
+        # type: (...) -> Tuple[np.ndarray, List[int]]
+        """Create X for prediction"""
+
+        trackers_as_states = self.prediction_states(trackers, domain)
+        X, true_lengths = self._featurize_states(trackers_as_states, domain)
+        return X, true_lengths
+
+    def persist(self, path):
+        featurizer_file = os.path.join(path, "featurizer.json")
+        utils.create_dir_for_file(featurizer_file)
+        with io.open(featurizer_file, 'w') as f:
+            f.write(str(jsonpickle.encode(self)))
+
+    @staticmethod
+    def load(path):
+        featurizer_file = os.path.join(path, "featurizer.json")
+        if os.path.isfile(featurizer_file):
+            with io.open(featurizer_file, 'r') as f:
+                _json = f.read()
+            return jsonpickle.decode(_json)
+        else:
+            logger.info("Couldn't load featurizer for policy. "
+                        "File '{}' doesn't exist. ".format(featurizer_file))
+            return None
+
+
+class FullDialogueFeaturizer(Featurizer):
+    def __init__(self, featurize_mechanism):
+        # type: (FeaturizeMechanism) -> None
+        super(FullDialogueFeaturizer, self).__init__(featurize_mechanism)
+
+        self.max_len = None
+
+    def _calculate_max_len(self, as_states):
+        self.max_len = 0
+        for states in as_states:
+            self.max_len = max(self.max_len, len(states))
+        logger.info("The longest dialogue has {} actions."
+                    "".format(self.max_len))
+
+    def _pad_states(self, states):
+        # pad up to max_len
+        if len(states) < self.max_len:
+            states += [None] * (self.max_len - len(states))
+
+        return states
+
+    def training_states_and_actions(
+            self,
+            trackers,  # type: List[DialogueStateTracker]
+            domain  # type: Domain
+    ):
+        # type: (...) -> Tuple[List[List[Dict]], List[List[Dict]], Dict]
+
+        trackers_as_states = []
+        trackers_as_actions = []
+        events_metadata = defaultdict(set)
+
+        for tracker in trackers:
+            states = domain.states_for_tracker_history(tracker)
+
+            delete_first_state = False
+            actions = []
+            for event in tracker._applied_events():
```

Comment:
I don't think we can only go through the applied events. e.g. consider:

```
## my nasty story
* greet
   - greet
   - undo
```
in this case the greet action will be reverted. while training the model we never generate the training sample to predict the action greet.

Reply:
we can go around here, by not adding the first action listen, instead of forgetting the last one

Reply:
It should be fine because during training we do not see explicitly `ActionReverted` or `UserUtteredReverted` because they are not part of the stories, so during training `_applied_events()=events` except for double action_listen due to augmentation.

During prediction phase, we go step by step so we should be able to predict `greet` and `undo` in this case. After undo we'll try to predict as if they never happen

Reply:
I append copy of the tracker to the list of trackers here:
https://github.com/RasaHQ/rasa_core/blob/1f991980cbb9387e9a5da9bcc69777c2fabb3264/rasa_core/training/generator.py#L247